docs/decisions/0063-vector-store-list-keys.md

```diff
@@ -0,0 +1,118 @@
+
+## Decision drivers
+1. We want to provide a way to list all keys in the vector store
+2. We want to provide this in a non-breaking way
+3. 
+
+## Considered Options
+
+- Option #1: add a new method to the VectorStoreRecordCollection interface
+- Option #2: add a new method to a new separate interface
+- Option #3: alter the behavior of the Get/GetBatch methods to allow getting all records (key=None or similar)
+
+## Option 1: add a new method to the VectorStoreRecordCollection interface
+
+We can add a new method to the VectorStoreRecordCollection interface to list all keys in the vector store. This method will return a list of keys.
+
+```csharp
+public interface IVectorStoreRecordCollection<TKey, TRecord> : IVectorizedSearch<TRecord>
+{
+    Task<TRecord?> GetAsync(TKey key, GetRecordOptions? options = default, CancellationToken cancellationToken = default);
+    IAsyncEnumerable<TRecord> GetBatchAsync(IEnumerable<TKey> keys, GetRecordOptions? options = default, CancellationToken cancellationToken = default);
+    ...
+    Task<IEnumerable<TKey>> ListKeysAsync(CancellationToken cancellationToken = default);
+}
+```
+Pros:
+- This is a simple and straightforward way to add the functionality
+Cons:
+- It is a breaking change (at least for csharp, python can do a default implementation that returns None)
+
+## Option 2: add a new method to a new separate interface
+
+We can add a separate interface to list all keys in the vector store. This method will return a list of keys.
+
+```csharp
+public interface IVectorStoreRecordCollectionListKeys<TKey, TRecord> : IVectorStoreRecordCollection<TKey, TRecord>
+{
+    Task<IEnumerable<TKey>> ListKeysAsync(CancellationToken cancellationToken = default);
+}
+```
+Pros:
+- This is a non-breaking change, each vector store record collection must choose to implement this interface.
+
+Cons:
+- This is a more complex way to add the functionality
+
+## Option 3: alter the behavior of the Get/GetBatch methods to allow getting all records (key=None or similar)
+
+This options alters the behavior of the Get methods to allow getting all records, this can be done by passing a special key value (e.g. None) to the Get method.
+
+```csharp
+
+public interface IVectorStoreRecordCollection<TKey, TRecord> : IVectorizedSearch<TRecord>
+{
+    Task<TRecord?> GetAsync(TKey? key, GetRecordOptions? options = default, CancellationToken cancellationToken = default);
+    IAsyncEnumerable<TRecord> GetBatchAsync(IEnumerable<TKey>? keys, GetRecordOptions? options = default, CancellationToken cancellationToken = default);
+}
+
+```
+Pros:
+- This allows the other options to be used in concert with getting all records
+- This is a non-breaking change
```

Comment:
It depends. For example, if someone already implemented our interface for their own connector and then we make the parameter nullable, it will raise warning `CS8767 - Nullability of reference types in type of parameter of doesn't match implicitly implemented member (possibly because of nullability attributes)`. For projects which handle warnings as errors, it will be a compilation error.